mlx_graphs/nn/conv/rel_conv.py

```diff
@@ -0,0 +1,354 @@
+from typing import Literal, Optional, get_args
+
+import mlx.core as mx
+from mlx import nn
+
+from mlx_graphs.nn.linear import Linear
+from mlx_graphs.nn.message_passing import MessagePassing
+from mlx_graphs.utils import degree, scatter
+
+MessageFunctions = Literal["transe", "distmult", "rotate"]
+AggegationFunctions = Literal["add", "mean", "pna"]
+
+
+class GeneralizedRelationalConv(MessagePassing):
+    """Generalized relational convolution layer from
+    `Neural Bellman-Ford Networks: A General Graph Neural Network Framework \
+      for Link Prediction <https://arxiv.org/abs/2106.06935>`_ paper.
+
+    Adopted from the PyG version from `here \
+    <https://github.com/KiddoZhu/NBFNet-PyG/blob/master/nbfnet/layers.py>`_.
+
+    Part of the Neural Bellman-Ford networks (NBFNet) holding
+    state-of-the-art in KG completion.
+    Works with multi-relational graphs where edge types are stored in `edge_labels`.
+    The message function composes node and relation vectors in three possible ways.
+    The expected behavior is to work with "labeling trick" graphs where one node
+    in the graph is labeled with a `query` vector, while rest are zeros.
+    Message passing is then done separately for each data point in the batch.
+    The input shape is expected to be [batch_size, num_nodes, input_dim]
+
+    Alternatively, the layer can work as a standard relational conv
+    with shapes [num_nodes, input_dim].
+
+    Note that this implementation materializes all edge messages and is O(E).
+    The complexity can be further reduced by adopting the O(V) `rspmm` C++ kernel
+    from the NBFNet-PyG repo to the MLX framework (not implemented here).
+
+    Args:
+        in_features_dim: input feature dimension (same for node and edge features)
+        out_features_dim: output node feature dimension
+        num_relations: number of unique relations in the graph
+        message_func: "transe" (sum), "distmult" (mult),
+            "rotate" (complex rotation). Default: ``distmult``
+        aggregate_func: "add", "mean", or "pna". Default: ``add``
+        layer_norm: whether to use layer norm
+            (often crucial to the performance). Default: ``True``
+        activation: non-linearity. Default: ``relu``
+        dependent: whether to use separate relation embedding matrix
+             ``False`` or build relations from the input relations ``True``
+        node_dim: for 3D batches, specified which dimension contains all nodes.
+            Default: ``0``
+
+    Example:
+
+    .. code-block:: python
+
+        import mlx.core as mx
+        import mlx.nn as nn
+        from mlx_graphs.nn import GeneralizedRelationalConv
+
```

Comment:
There's still the errors in this example, can you please run it locally so that  we're sure it's reproducible? :)